src/core/ext/transport/chttp2/transport/hpack_table.h

```diff
@@ -46,41 +46,65 @@
+    return (bytes + GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD - 1) /
+           GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+  }
+  static constexpr uint32_t InitialCapacity =
+      (/*bytes=*/GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE +
+       GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD - 1) /
+      GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+
+  grpc_chttp2_hptbl() = default;
+
   /* the first used entry in ents */
-  uint32_t first_ent;
+  uint32_t first_ent = 0;
   /* how many entries are in the table */
-  uint32_t num_ents;
+  uint32_t num_ents = 0;
   /* the amount of memory used by the table, according to the hpack algorithm */
-  uint32_t mem_used;
+  uint32_t mem_used = 0;
   /* the max memory allowed to be used by the table, according to the hpack
      algorithm */
-  uint32_t max_bytes;
+  uint32_t max_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* the currently agreed size of the table, according to the hpack algorithm */
-  uint32_t current_table_bytes;
+  uint32_t current_table_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* Maximum number of entries we could possibly fit in the table, given defined
      overheads */
-  uint32_t max_entries;
+  uint32_t max_entries = InitialCapacity;
   /* Number of entries allocated in ents */
-  uint32_t cap_entries;
+  uint32_t cap_entries = InitialCapacity;
   /* a circular buffer of headers - this is stored in the opposite order to
      what hpack specifies, in order to simplify table management a little...
      meaning lookups need to SUBTRACT from the end position */
-  grpc_mdelem* ents;
-  grpc_mdelem static_ents[GRPC_CHTTP2_LAST_STATIC_ENTRY];
-} grpc_chttp2_hptbl;
+  grpc_mdelem* ents = nullptr;
+};
 
 /* initialize a hpack table */
-void grpc_chttp2_hptbl_init(grpc_chttp2_hptbl* tbl);
+inline void grpc_chttp2_hptbl_init(grpc_chttp2_hptbl* tbl) {
+  new (tbl) grpc_chttp2_hptbl();
+  constexpr uint32_t AllocSize =
+      sizeof(*tbl->ents) * grpc_chttp2_hptbl::InitialCapacity;
+  tbl->ents = static_cast<grpc_mdelem*>(gpr_malloc(AllocSize));
+  memset(tbl->ents, 0, AllocSize);
+}
+
 void grpc_chttp2_hptbl_destroy(grpc_chttp2_hptbl* tbl);
 void grpc_chttp2_hptbl_set_max_bytes(grpc_chttp2_hptbl* tbl,
                                      uint32_t max_bytes);
 grpc_error* grpc_chttp2_hptbl_set_current_table_size(grpc_chttp2_hptbl* tbl,
                                                      uint32_t bytes);
 
 /* lookup a table entry based on its hpack index */
-grpc_mdelem grpc_chttp2_hptbl_lookup(const grpc_chttp2_hptbl* tbl,
-                                     uint32_t index);
+grpc_mdelem grpc_chttp2_hptbl_lookup_slow(const grpc_chttp2_hptbl* tbl,
+                                          uint32_t tbl_index);
+inline grpc_mdelem grpc_chttp2_hptbl_lookup(const grpc_chttp2_hptbl* tbl,
+                                            uint32_t index) {
+  /* Static table comes first, just return an entry from it */
+  return index <= GRPC_CHTTP2_LAST_STATIC_ENTRY
+             ? grpc_static_mdelem_manifested[index - 1]
```

Comment:
Does this mean that the first GRPC_CHTTP_LAST_STATIC_ENTRY elements of grpc_static_mdelem_manifested must adhere to the hpack spec for static hpack table ordering?  If so, please add a comment to the static_metadata files as well as to gen_static_metadata.py so that someone doesn't accidentally change this in the future.

Reply:
Will do.

Reply:
Added comments making this assumption explicit and indicating that if it is to be changed, it must be fixed in both core and CH2.